### PR TITLE
Лахтачев -- Лабораторная работа #2 

### PR DESCRIPTION
**Pull request by Yaroslav-Lahtachev**

### CODE REVIEW

**Comment by Yaroslav-Lahtachev** [comment]
@Selllavi @ViolettaShirokikh 
hi colleagues, can you review second lab?

File: code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    private StringProperty input = new SimpleStringProperty();
+    private StringProperty output = new SimpleStringProperty();
+    private StringProperty error = new SimpleStringProperty();
+    private StringProperty convert = new SimpleStringProperty();
+    private BooleanProperty btnDisabled = new SimpleBooleanProperty();
+    private List<NumConverter> converters = new ArrayList<NumConverter>();
+    private int selector = 0;
+
+    private void prepareConverters() {
+        ArabicToRomanConverter converterA = new ArabicToRomanConverter();
```
Comment (Selllavi) [inline code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java]: converterC...ConverterD...ConverterZ...ConverterAB………...

File: code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        RomanToArabicConverter converterB = new RomanToArabicConverter();
+        converters.add(converterA);
+        converters.add(converterB);
+    }
+
+    public ViewModel() {
+        refresh();
+        convert.set("Arabic to Romain");
```
Comment (Selllavi) [inline code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java]: Romain is incorrect type of number.
They are Roman numbers)))

File: code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+    public void convert() {
+        String value = input.get();
+        NumConverter converter = converters.get(selector);
+        String result = converter.convert(value);
+        output.set(result);
+    }
+
+    public void swap() {
+        if (selector == 0) {
```
Comment (Selllavi) [inline code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java]: It looks like magical number, maybe will be better to create ENUM with Options?

File: code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            selector = 1;
+            convert.set("Romain to Arabic");
```
Comment (Selllavi) [inline code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java]: And add to this enum labels with "Roman to Arabic" or "Arabic to Roman". Do not copy-paste these strings everywhere.

File: code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            selector = 0;
+            convert.set("Arabic to Romain");
+            refresh();
+        }
+    }
+
+    public void refresh() {
+        input.set("");
+        output.set("");
+        error.set("");
```
Comment (Selllavi) [inline code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java:44]: Maybe convert.set also should be done in refresh()?

File: code/arabic-roman-converter/Model/src/main/java/ru/unn/agile/arabicromanconverter/model/ArabicToRomanConverter.java [hunk 6]
```diff
@@ -9,9 +9,17 @@
     private static final int MIN_ARABIC_VALUE = 1;
     private static final int MAX_ARABIC_VALUE = 3999;
 
-    public ArabicToRomanConverter() { };
+    public ArabicToRomanConverter() {
+    }
+
+    public boolean validate(final String num) {
```
Comment (Selllavi) [inline code/arabic-roman-converter/Model/src/main/java/ru/unn/agile/arabicromanconverter/model/ArabicToRomanConverter.java]: why everywhere num instead of number?

File: code/arabic-roman-converter/ViewModel/src/test/java/ru/unn/agile/arabicromanconverter/viewmodel/ArabicRomanConverterViewModelTests.java [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        assertEquals(true, viewModel.isConvertButtonDisabled().get());
+    }
+
+    @Test
+    public void canConvertWhileNonValidationErrorRomanToArabic() {
+        viewModel.swap();
+        viewModel.getInput().set("I");
+
+        assertEquals(false, viewModel.isConvertButtonDisabled().get());
+    }
```
Comment (Selllavi) [inline code/arabic-roman-converter/ViewModel/src/test/java/ru/unn/agile/arabicromanconverter/viewmodel/ArabicRomanConverterViewModelTests.java:146]: Add some tests to check behaviour if user set incorrect number("BUGAGASHEN'KA") and number that doesn't exist in Roman numbers. (-1 for example)

File: code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java (hunk 1 above)
Comment (Yaroslav-Lahtachev) [reply code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java]: fixed after enum added

File: code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java (hunk 3 above)
Comment (Yaroslav-Lahtachev) [reply code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java]: fixed after enum added

File: code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java (hunk 4 above)
Comment (Yaroslav-Lahtachev) [reply code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java]: fixed after enum added

File: code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java (hunk 5 above)
Comment (Yaroslav-Lahtachev) [reply code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java:44]: fixed after enum added

File: code/arabic-roman-converter/Model/src/main/java/ru/unn/agile/arabicromanconverter/model/ArabicToRomanConverter.java (hunk 6 above)
Comment (Yaroslav-Lahtachev) [reply code/arabic-roman-converter/Model/src/main/java/ru/unn/agile/arabicromanconverter/model/ArabicToRomanConverter.java]: oh God.... ok...

File: code/arabic-roman-converter/ViewModel/src/test/java/ru/unn/agile/arabicromanconverter/viewmodel/ArabicRomanConverterViewModelTests.java (hunk 7 above)
Comment (Yaroslav-Lahtachev) [reply code/arabic-roman-converter/ViewModel/src/test/java/ru/unn/agile/arabicromanconverter/viewmodel/ArabicRomanConverterViewModelTests.java:146]: Add test for boundary values

File: code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    public StringProperty getOutput() {
+        return output;
+    }
+
+    public StringProperty getConvert() {
+        return convert;
+    }
+
+
+    public StringProperty getError() {
```
Comment (ViolettaShirokikh) [inline code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java]: remove one empty line

File: code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        btnDisabled.set(newValue.isEmpty() || !isValid);
+        output.set("");
+    }
+}
+
+enum ConverterType {
+    ARABICTOROMAN("Arabic to Roman", new ArabicToRomanConverter(), "Insert correct arabic number"),
+    ROMANTOARABIC("Roman to Arabic", new RomanToArabicConverter(), "Insert correct roman number");
+
+    private final String label;
```
Comment (ViolettaShirokikh) [inline code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java]: What is label? Name should be meaningful.

**Comment by Selllavi** [comment]
👍 thanks

File: code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java (hunk 9 above)
Comment (Selllavi) [reply code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java]: But it is quite meaningful for me. It is label, what do you suggest to Yaroslav? :D

Comment (ViolettaShirokikh) [reply code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java]: "lblConvertingType" :D

Comment (Yaroslav-Lahtachev) [reply code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java]: done

File: code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java (hunk 8 above)
Comment (Yaroslav-Lahtachev) [reply code/arabic-roman-converter/ViewModel/src/main/java/ru/unn/agile/arabicromanconverter/viewmodel/ViewModel.java]: done

**Comment by Yaroslav-Lahtachev** [comment]
@ViolettaShirokikh please, review again)

**Comment by ViolettaShirokikh** [comment]
:+1:

File: code/arabic-roman-converter/Model/src/main/java/ru/unn/agile/arabicromanconverter/model/NumConverter.java [hunk 10]
```diff
@@ -0,0 +1,7 @@
+package ru.unn.agile.arabicromanconverter.model;
+
+public interface NumConverter {
+    String convert(String number);
```
Comment (vgvoleg) [inline code/arabic-roman-converter/Model/src/main/java/ru/unn/agile/arabicromanconverter/model/NumConverter.java]: Объясните пожалуйста, почему вместо конвертации стрингового входа в число для передачи конвертеру, было принято решение сломать апи конвертера?

**Review by vgvoleg — CHANGES_REQUESTED** [review]

File: code/arabic-roman-converter/Model/src/main/java/ru/unn/agile/arabicromanconverter/model/NumConverter.java (hunk 10 above)
Comment (Yaroslav-Lahtachev) [reply code/arabic-roman-converter/Model/src/main/java/ru/unn/agile/arabicromanconverter/model/NumConverter.java]: Для унификации, апи внешне работает полностью со стрингом.

Comment (vgvoleg) [reply code/arabic-roman-converter/Model/src/main/java/ru/unn/agile/arabicromanconverter/model/NumConverter.java]: Это не унификация, ты сломал старое апи, которое успешно справлялось с числами. Ничего не мешает расширить интерфейс для работы со строками, чтобы не ломать старый метод, работающий с числами. И не придется переписывать тесты с (0) на ("0")

File: code/arabic-roman-converter/Model/src/main/java/ru/unn/agile/arabicromanconverter/model/ArabicToRomanConverter.java [hunk 11]
```diff
@@ -9,9 +9,17 @@
     private static final int MIN_ARABIC_VALUE = 1;
     private static final int MAX_ARABIC_VALUE = 3999;
 
-    public ArabicToRomanConverter() { };
+    public ArabicToRomanConverter() {
```
Comment (vgvoleg) [inline code/arabic-roman-converter/Model/src/main/java/ru/unn/agile/arabicromanconverter/model/ArabicToRomanConverter.java:12]: почему вообще методы у этого класса не статические, а конструктор не приватный? это и ко второму конвертеру относится

**Review by vgvoleg — CHANGES_REQUESTED** [review]

File: code/arabic-roman-converter/Model/src/main/java/ru/unn/agile/arabicromanconverter/model/NumConverter.java (hunk 10 above)
Comment (Yaroslav-Lahtachev) [reply code/arabic-roman-converter/Model/src/main/java/ru/unn/agile/arabicromanconverter/model/NumConverter.java]: вернул все как было

File: code/arabic-roman-converter/Model/src/main/java/ru/unn/agile/arabicromanconverter/model/ArabicToRomanConverter.java (hunk 11 above)
Comment (Yaroslav-Lahtachev) [reply code/arabic-roman-converter/Model/src/main/java/ru/unn/agile/arabicromanconverter/model/ArabicToRomanConverter.java:12]: Наверное по той же причине почему я откатил и свои изменения? Потому что придется переписать тесты.
